pkg/controller/ipam/metrics.go

```diff
@@ -0,0 +1,150 @@
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+	ipPoolUsageUsed = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      metricNamespaceAntrea,
+			Subsystem:      metricSubsystemIPAM,
+			Name:           "ippool_usage_used",
+			Help:           "Number of allocated IP addresses for the address pool.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+)
+
+func NewAntreaIPAMMetricsHandler(crdClient versioned.Interface, crdInformerFactory crdinformers.SharedInformerFactory) *AntreaIPAMMetricsHandler {
+	legacyregistry.MustRegister(ipPoolUsageTotal)
+	legacyregistry.MustRegister(ipPoolUsageUsed)
+
+	ipPoolInformer := crdInformerFactory.Crd().V1alpha2().IPPools()
+	mh := &AntreaIPAMMetricsHandler{
+		crdClient:      crdClient,
+		ipPoolInformer: ipPoolInformer,
+		ipPoolLister:   ipPoolInformer.Lister()}
+	return mh
+}
+
+func (c *AntreaIPAMMetricsHandler) Run(stopCh <-chan struct{}) {
+	c.ipPoolInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.createHandler,
+		UpdateFunc: c.updateHandler,
+		DeleteFunc: c.deleteHandler,
+	}, 0)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformer.Informer().HasSynced) {
+		return
+	}
+	<-stopCh
+}
+
+func (c *AntreaIPAMMetricsHandler) createHandler(obj interface{}) {
+	c.updateIPPoolCounters(obj)
+}
+
+func (c *AntreaIPAMMetricsHandler) updateHandler(oldObj, newObj interface{}) {
+	c.updateIPPoolCounters(newObj)
+}
+
+func (c *AntreaIPAMMetricsHandler) deleteHandler(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	if !ipPoolUsageTotal.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+	if !ipPoolUsageUsed.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+}
+
+func (c *AntreaIPAMMetricsHandler) updateIPPoolCounters(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPool.Name, c.crdClient)
+
+	if err != nil {
+		klog.Warningf("Failed to initialize allocator for IPPool %s", ipPool.Name)
+	}
+
+	// Total is fetched from allocator as here are trapped changes to CRD, e.g addition of new IPRange
+	total := allocator.Total()
+
+	// Used is gathered from IP allocation status within the CRD - as it can be set by each one of the agents
+	used := len(ipPool.Status.IPAddresses)
+
+	// Set Prometheus metrics
+	ipPoolUsageTotal.WithLabelValues(ipPool.Name).Set(float64(total))
+	ipPoolUsageUsed.WithLabelValues(ipPool.Name).Set(float64(used))
+
+	// Update the status within the CRD
+	pool, err := c.ipPoolLister.Get(ipPool.Name)
+	if err != nil {
+		klog.Warningf("Failed to retrieve pool %s from CRD", ipPool.Name)
+		return
+	}
+
+	// If update has no effect, exit
+	if pool.Status.Usage.Used == used && pool.Status.Usage.Total == total {
+		return
+	}
+	pool.Status.Usage.Used = used
```

Comment:
Would it make more sense to update these counters along with other Status updates (when addresses are actually allocated/released), rather then asynchronously here? This would also save the need for an extra CRD update.

Reply:
I tried this actually, here is the problems though:
If I update the total number in the other status update, it won't catch IPPool creation or updates when an IPRange is added. So for the total it makes sense to use an informer.
As for updating the used count, I could indeed place it where the other updates are done. It will clutter the code though a bit.
Another option is using a mutating webhook, I didn't try it though.
What do you think?

Reply:
Yes, this makes sense to me, as long as we perform the update here only if totals are changed.

Reply:
@tnqn the change does expose Prometheus metrics as well as the stats via CRD. As this PR has been around for while, I can't recall why it's so. I can remove this.
This change has to be revised, maybe discard the Prometheus code as you suggest.